### PR TITLE
feat: migrate `cross-fetch` to `ofetch`

### PR DESCRIPTION
**Pull request by wa0x6e**
This PR is migrating the fetch library from `cross-fetch` to `ofetch`.

This new library add supports for :
- `keep-alive` connections on nodejs (need to add `FETCH_KEEP_ALIVE=true` to .env var).
- automatic response parsing (keeps the response parsing code shorter)
- timeout (set to 10s, although pineapple usually respond within 1s)

Code also refactored to:
- return a JSON-RPC error on generic network errors (504, 404, etc ...). All errors returned by this lib now follow the same format.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,6 +50,10 @@ On error, it will return an error object like this:
 }
 ```
 
+### Notes
+
+- On NodeJS, you can set the env variable `FETCH_KEEP_ALIVE` to `true` to enable keep aliuve connections.
```
Comment (Todmy) [inline README.md]:
```suggestion
- On NodeJS, you can set the env variable `FETCH_KEEP_ALIVE` to `true` to enable keep-alive connections.
```

File: src/index.ts [hunk 2]
```diff
@@ -1,29 +1,37 @@
...
-  const init = { method: 'POST', body };
-  const res = await fetch(url, init);
-  const content = await res.json();
-  return content.result || { error: content.error };
+  const init = { method: 'POST', body, timeout };
+
+  return sendRequest(url, init);
+}
+
+async function sendRequest(url: string, init: any) {
```
Comment (Todmy) [inline src/index.ts]:
```suggestion
async function sendRequest(url: string, body: any) {
```

File: src/index.ts [hunk 3]
```diff
@@ -1,29 +1,37 @@
...
+    timeout
   };
-  const res = await fetch(url, init);
-  const content = await res.json();
-  return content.result || { error: content.error };
+
+  return sendRequest(url, init);
 }
 
 export async function upload(body: any, url = `${PINEAPPLE_URL}/upload`) {
```
Comment (Todmy) [inline src/index.ts]: you don't need `async` here

File: src/index.ts [hunk 4]
```diff
@@ -1,29 +1,37 @@
...
+  const init = { method: 'POST', body, timeout };
+
+  return sendRequest(url, init);
+}
+
+async function sendRequest(url: string, init: any) {
+  try {
+    return (await fetch(url, init)).result;
+  } catch (e: any) {
+    return { error: e.data?.error || { code: e.status, message: e.statusText } };
```
Comment (Todmy) [inline src/index.ts]: it should throw the error due to usage [here](https://github.com/snapshot-labs/snapshot/blob/94d82953023f9e90a8a1027490711db2b71736ea/src/components/InputUploadImage.vue#L25), [here](https://github.com/snapshot-labs/snapshot-sequencer/blob/01b5c1ad78f56704e17eace3bc2e56f101442a0c/src/ingestor.ts#L195) and [here](https://github.com/snapshot-labs/sx-ui/blob/8412e229e838cc45684f4b81d629f26fa758a2ef/src/helpers/utils.ts#L345)

**Review by Todmy — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 4 above)
Comment (Todmy) [reply src/index.ts]: [exception](https://github.com/snapshot-labs/snapshot/blob/94d82953023f9e90a8a1027490711db2b71736ea/src/composables/useImageUpload.ts#L44)

Comment (wa0x6e) [reply src/index.ts]: Updated to return a `Promise.rejection` on error. This will handle all the cases where we use try/catch when using `pin`

Comment (wa0x6e) [reply src/index.ts]: For the exception case (https://github.com/snapshot-labs/snapshot/blob/94d82953023f9e90a8a1027490711db2b71736ea/src/composables/useImageUpload.ts#L44), it'll need some refactoring, as the new implementation will always return a `Promise.rejection` on error, and always in the form of a JSON-RPC object.

File: README.md [hunk 5]
```diff
@@ -50,6 +50,10 @@ On error, it will return an error object like this:
 }
 ```
 
+### Notes
+
+- On NodeJS, you can set the env variable `FETCH_KEEP_ALIVE` to `true` to enable keep-alive connections.
```
Comment (ChaituVR) [inline README.md:55]: We are not using this anywhere? 🤔

**Review by ChaituVR — DISMISSED** [review]

File: README.md (hunk 5 above)
Comment (ChaituVR) [inline README.md:55]: we have to set this ENV on all projects again? 🙈 maybe we can set it true from code itself using dotenv?? 

https://github.com/motdotla/dotenv#processenv

Comment (wa0x6e) [reply README.md:55]: We can set it inside .env yes

Comment (ChaituVR) [reply README.md:55]: No i mean without having to change env on digital ocean

Comment (wa0x6e) [reply README.md:55]: Yes, by adding it to each project .env file. If they’re not using dotenv, still need to change do env

Comment (ChaituVR) [reply README.md:55]: Cool, will change in this PR itself?

Comment (wa0x6e) [reply README.md:55]: No, the env var has to be added to the service using this package

**Review by Todmy — APPROVED** [review]
utACK

**Pull request merged**